File: examples/src/main/java/com/edgedb/examples/Transactions.java
package com.edgedb.examples;

import com.edgedb.driver.EdgeDBClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletionStage;

public final class Transactions implements Example {
    private static final Logger logger = LoggerFactory.getLogger(Transactions.class);

    @Override
    public CompletionStage<Void> run(EdgeDBClient client) {
        return client.transaction(tx -> {
            logger.info("In transaction");
            return tx.queryRequiredSingle(String.class, "select 'Result from Transaction'");
        }).thenAccept(result -> {
            logger.info("Result from transaction: {}", result);
        });
    }
}

File: examples/kotlin-examples/src/main/kotlin/com/edgedb/examples/Transactions.kt
package com.edgedb.examples

import com.edgedb.driver.GelClientPool
import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory

class Transactions : Example {
    companion object {
        private val logger = LoggerFactory.getLogger(Transactions::class.java)
    }

    override suspend fun runAsync(clientPool: GelClientPool) {
        // verify we can run transactions
        if (!client.supportsTransactions()) {
            logger.info("Skipping transactions, client type {} doesn't support it", client.clientType)
            return
        }

        val transactionResult = client.transaction { tx ->
            tx.queryRequiredSingle(String::class.java, "SELECT 'Hello from transaction!'")
        }.await()

        logger.info("Transaction result: {}", transactionResult)
    }
}